docs/bypass.rst

```diff
@@ -0,0 +1,68 @@
+.. _bypass:
+
+Bypassing Blocking Locks
+========================
+
+`pglock.bypass` ensures that any blocking locks of a piece of code will be terminated.
+A background thread carries out termination of blocking statements on a configurable
+periodic interval.
+
+Basic Usage
+-----------
+
+For example, below we've ensured that our function kills any blocking activity
+during its execution:
+
+.. code-block:: python
+
+    import pglock
+
+    @pglock.bypass()
+    def my_important_function():
+        # If any statements block this function, they will be terminated.
+
+By default, the background thread operations on a one-second interval, but this can
+be configured with the ``interval`` argument:
+
+.. code-block:: python
+
+    @pglock.bypass(interval=pglock.timedelta(seconds=5))
+    def my_important_function():
+        # Every five seconds, blocking activity will be terminated
+
+Side Effects
+------------
+
+The ``side_effect`` argument can be used to configure the function executed
+by the background worker. It defaults to ``pglock.TERMINATE``.
+
+The side effect is a function that is supplied a `BlockedPGLock` queryset that
+matches the process ID of the bypassed code. It must return a list of process IDs
+that were handled. For example, the ``pglock.TERMINATE`` side effect looks like this:
+
+.. code-block:: python
+
+    def terminate_side_effect(blocking_locks):
+        return blocking_locks.terminate_blocking_activity()
+
+If you don't want to terminate processes, you can use ``pglock.bypass(side_effect=pglock.CANCEL)``
+to cancel blocking activity; however,
+keep in mind that not all queries can be canceled.
+
+Other Arguments
+---------------
+
+Here are the other arguments of `pglock.bypass` for more advanced use cases:
+
+* **timeout**: Set the lock timeout. This is equivalent to calling
+  ``with pglock.bypass(), pglock.timeout(...):``. This can be used
+  as an additional safety net in the case the background worker cannot
+  properly terminate blocking activity.
+* **retries**: Retry the side effect this many times.
+  No retries are performed by default.
+  For example, other blocking row-level locks
+  may arise immediately after other locks have been cleared. Using ``retries`` ensures another
```

Comment:
A little confused by this description. The scenario you describe where a lock can appear again immediately after terminating is a common one, but I can't see why we'd try to solve it by passing in a finite 'retries' value. For example, if I was the end user, passing in e.g. retries=2 would feel the same as not passing in a retry at all. If I can't be sure that the lock won't come back after a single terminate action, why can I be any more sure about it not coming back after 3 terminate actions?

It seems to be that this scenario would be better solve by something like a `continual` argument which just terminates in a loop?

Reply:
Yeah I think I'm on the same page and understand what retries does. It was just more if I was a dev using it for the behaviour you describe (i.e. wanting to not wait for the period to finish before killing more PIDs), I'd never know what finite number to choose. It's obvious in your example because we laid out the queries before hand, but of course that wouldn't really ever be a real world scenario. That's why I suggested just having a 'continual' option which just kills constantly, which is what I'd want most often in an emergency.